Zotero/Controllers/OpenItemsController.swift

```diff
@@ -0,0 +1,481 @@
+    private var observableBySessionIdentifier: [String: PublishSubject<[Item]>] = [:]
+    private let disposeBag: DisposeBag
+
+    // MARK: Object Lifecycle
+    init(dbStorage: DbStorage, fileStorage: FileStorage) {
+        self.dbStorage = dbStorage
+        self.fileStorage = fileStorage
+        disposeBag = DisposeBag()
+    }
+    
+    // MARK: Actions
+    func observable(for sessionIdentifier: String) -> PublishSubject<[Item]> {
+        if let observable = observableBySessionIdentifier[sessionIdentifier] {
+            return observable
+        }
+        let observable = PublishSubject<[Item]>()
+        observableBySessionIdentifier[sessionIdentifier] = observable
+        return observable
+    }
+
+    func getItems(for sessionIdentifier: String) -> [Item] {
+        itemsBySessionIdentifier[sessionIdentifier, default: []]
+    }
+
+    func setItems(_ items: [Item], for sessionIdentifier: String, validate: Bool) {
+        DDLogInfo("OpenItemsController: setting items \(items) for \(sessionIdentifier)")
+        let existingItems = getItems(for: sessionIdentifier)
+        let newItems = validate ? filterValidItems(items) : items
+        guard newItems != existingItems else { return }
+        // Invalidate previous observer first.
+        itemsTokenBySessionIdentifier[sessionIdentifier]?.invalidate()
+        itemsTokenBySessionIdentifier[sessionIdentifier] = nil
+        // Update itemsBySessionIdentifier.
+        itemsBySessionIdentifier[sessionIdentifier] = newItems
+        // Update sessionIdentifierByItemKind. Recompute for all session identifier, to remove any closed items.
+        var newSessionIdentifierByItemKind: [Item.Kind: String] = [:]
+        itemsBySessionIdentifier.forEach { (sessionIdentifier, items) in
+            items.forEach { item in
+                newSessionIdentifierByItemKind[item.kind] = sessionIdentifier
+            }
+        }
+        sessionIdentifierByItemKind = newSessionIdentifierByItemKind
+        // Register observer for newly set items.
+        itemsTokenBySessionIdentifier[sessionIdentifier] = registerObserver(for: newItems)
+        observable(for: sessionIdentifier).on(.next(newItems))
+
+        func registerObserver(for items: [Item]) -> NotificationToken? {
+            var token: NotificationToken?
+            var keysByLibraryIdentifier: [LibraryIdentifier: Set<String>] = [:]
+            for item in items {
+                let libraryId = item.kind.libraryId
+                let key = item.kind.key
+                var keys = keysByLibraryIdentifier[libraryId, default: .init()]
+                keys.insert(key)
+                keysByLibraryIdentifier[libraryId] = keys
+            }
+            do {
+                try dbStorage.perform(on: .main) { coordinator in
+                    let objects = try coordinator.perform(request: ReadItemsWithKeysFromMultipleLibrariesDbRequest(keysByLibraryIdentifier: keysByLibraryIdentifier))
```

Comment:
Same as above, this doesn't need to happen inside closure with coordinator. You can just get `objects` directly from `dbStorage` and observe them.

Reply:
Check https://github.com/zotero/zotero-ios/pull/829/commits/ae2376422f3f11d32545767dac3d7b35913fac7a